### PR TITLE
[DEV-2349] Add `DefinitionHashExtractor` to extract feature definition hash

### PR DESCRIPTION
**Pull request by gee-senbong**
## Description

This PR introduces `DefinitionHashExtractor` class that will be used to generate feature definition hash on saved feature graph.

## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [ ] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [ ] I have written tests for the changes made.
- [ ] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [ ] I have labeled my Pull Request correctly

### CODE REVIEW

File: featurebyte/query_graph/node/mixin.py [hunk 1]
```diff
@@ -225,3 +226,29 @@ def _derive_node_operation_info(
...
+        # remap keys
+        remapped_keys = []
+        for key in self.parameters.keys:
+            remapped_key = input_node_column_remap.get(key, key)
+            remapped_keys.append(InColumnStr(remapped_key))
+        remapped_node.parameters.keys = remapped_keys
+
+        # remap parent and value_by
+        if remapped_node.parameters.parent in input_node_column_remap:
+            remapped_parent = input_node_column_remap[remapped_node.parameters.parent]
```
Comment (YS-L) [inline featurebyte/query_graph/node/mixin.py:249]: Does this have to use `get` like above for keys?

File: featurebyte/query_graph/node/generic.py [hunk 2]
```diff
@@ -1227,6 +1436,61 @@ def _derive_sdk_code(
...
+            remapped_right_output_columns,
+        )
+        remapped_node.parameters.left_input_columns = left_in_cols
+        remapped_node.parameters.left_output_columns = left_out_cols  # type: ignore
+        remapped_node.parameters.right_input_columns = right_in_cols
+        remapped_node.parameters.right_output_columns = right_out_cols  # type: ignore
+
+        # reset metadata
+        remapped_node.parameters.metadata = None
+        remapped_node.parameters.scd_parameters = None
```
Comment (YS-L) [inline featurebyte/query_graph/node/generic.py:1491]: Is it safe to reset the scd_parameters here?

Comment (gee-senbong) [reply featurebyte/query_graph/node/generic.py:1491]: It is possible to construct the same graph structure with this `scd_parameters` set differently? If yes, we should not reset these parameters.

File: featurebyte/query_graph/transform/definition.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        # create the node with column name remap
+        input_node_hashes = [
+            global_state.graph.node_name_to_ref[input_node.name] for input_node in inputs
+        ]
+        column_remapped_node, column_name_remap = node.convert_to_column_remapped_node(
+            input_node_hashes=input_node_hashes,
+            input_node_column_remaps=[
+                global_state.node_name_to_column_name_remap[input_node_name]
+                for input_node_name in self.graph.get_input_node_names(node)
+            ],
```
Comment (YS-L) [inline featurebyte/query_graph/transform/definition.py:112]: Does the ordering of these two lists need to correspond to each other? The first one has sorting applied.

Comment (gee-senbong) [reply featurebyte/query_graph/transform/definition.py:112]: This is a very good catch! We should sort the second one too 🙏

**Review by YS-L — APPROVED** [review]

**Pull request merged**